pypeclub/pype
# Feature/PYPE-762 multi root

### Integration of multi root
- At the time of PR creation it is more draft and **NOT final version**!!!
- There are many parts I'm sure I forgot about, but need to test it at this stage to know where exatly these parts are.
- Please let in comments what do you think. (I'll let you what I think).
    - Please make comments inside the code to easier see what are you talking about.

### Changes
- `~/pype-config/system/` was removed
    - `storage.json` has changed to `roots.json` in anatomy
    - `core.json` was for setting `PYPE_STUDIO_CORE_PATH/MOUNT` which as was found is not necessary
- `Storage` class was removed from pypeapp
- for backwards compatibility there are 2 possiblities.
1.) set default `roots.json` without anatomy keys same as `storage.json` was.
2.) copy exact the same json as mentioned in (1) into each project overrides.
```
EXAMPLE OF BACKWARDS COMPATIBLE roots.json
{
    "windows": "P:/projects/publish",
    "linux": "/mnt/share/projects/publish",
    "darwin": "/darwin/path/publish"
}
```
- Ftrack action `Prepare Project`  asks for filling of each root key set in `default.yaml`
    - there is logic how it is pre-filled which I can explain if anybody asks
- Anatomy class is now a wrapper for `Templates` and `Roots`.
    - If you are using anatomy for creating paths, you should not care about roots.
        - Not really true inside `avalon`, but most use cases of avalon are inside host where project is already specified so no worries now.
- this PR requires to NEVER EVER use environments for getting ROOT in code, always MUST use `Anatomy`, `Roots` object for that.
- Roots object has `find_root_template_from_path` which can return path with `{root[root type]}` if project roots are matching entered path.
    - similar feature should be added to filled template by anatomy. Filled template should have `rootless` attribute where path without root key will be stored. With this it may be easier to achieve partially filled paths.

### Testing
- check **TODO/Question** below
- create new project (if you don't want to break your other project)
- copy `roots.json` to project override or use prepare action as explained above
- run application
- publish
- load
- publish to farm
- watch errors

### TODOs/Questions/Issues
- publish paths are not correct on farm (I am 99% sure)
    - this is more question than TODO, how to do it
    - I had idea to have partially filled paths (without root) until integration happens
        - That is hard to do, each host has many, many parts where path are created
- Ftrack action `Create project folders` do not use Anatomy for creating folders right now and folders are created for each root defined in project's template, which may cause issues in future.
- `thumbnails_root` was kept as it is because we exect to store in one place. But it easy to replace `{thumbnail_root}` with key anybody wants
